core/src/oidc_auth.rs

```diff
@@ -175,3 +182,49 @@ pub async fn do_auth_flow(params: CallbackContext) -> Result<IdpServerResponse,
         refresh_token,
     })
 }
+
+pub async fn do_refresh(params: CallbackContext) -> Result<IdpServerResponse, Error> {
+    let idp_info = params.idp_info.ok_or(Error::NoIdpServerInfo)?;
+    let client_id = idp_info.client_id;
+    let issuer_uri = IssuerUrl::new(idp_info.issuer).map_err(|e| Error::Other(e.to_string()))?;
+    if issuer_uri.url().scheme() != "https" {
+        return Err(Error::IssuerUriMustBeHttps);
+    }
+
+    // Use OpenID Connect Discovery to fetch the provider metadata.
+    let provider_metadata = CoreProviderMetadata::discover_async(issuer_uri, async_http_client)
+        .await
+        .map_err(|e| Error::Other(e.to_string()))?;
+
+    // Create an OpenID Connect client by specifying the client ID,
+    // authorization URL and token URL.
+    let client = CoreClient::from_provider_metadata(
+        provider_metadata,
+        ClientId::new(client_id),
+        None,
+    );
+
+    // This must never be called without a refresh token, but we return this error anyway
```

Comment:
Not sure I understand the comment completely. 
I agree it is good to protect our-self here and I believe this is what you imply, correct?

A short comment on top of the function, describing what it is used for (I know, it's mostly self explanatory, but it won't hurt) and stating what fields must be present is also a good idea. 
Maybe for `do_auth_flow` too, since idp_info is an option but it is actually required.

Reply:
Ah, yes, the comment would make a lot more sense if I'd already implemented the next ticket. This function will only ever be called when there is a refresh_token, but I decided to be defensive anyway